### PR TITLE
Refactor contracts class and Implement redaction

### PR DESCRIPTION
**Pull request by silasdavis**
**note: is isn't quite as bad as the LOC imply some fairly mechanical refactoring end up hitting most lines in contract-controller.js. Suggest local review and you can get a somewhat usable diff from `git diff origin/master:api/controllers/contracts-controller.js HEAD:api/contracts.js`**

In testing redaction in API and found various issues with vent synchronisation within contracts-controller. Synchronising on every call worked but was verging on unacceptably slow for some sequences requiring many writes (there is more scope to parallelise Burrow calls by just awaiting later, but that is for another day). In the end I refactored `contracts-controller` into `Contracts` client class with some regex assistance that provides better semantics/syntax for batching together contracts calls. The design is for it to be safe (i.e. sync) by default but allow the caller to avoid unnecessary synchronisation between a series of calls. This also looks to me to be a better structure to evolve into a `AN.js` and we can use the structure here to reimplement existing batch calls as single Burrow transactions or to call WASM backends (not something I was particularly aiming for though).  So we have:

I've kept these changes separate. If we are happy to merge this it might be better @j-h-scheufen for you to work you changes to my redaction contract code on a later commit based on these changes.

## Contracts refactor

This provides finer-grained control over waiting for vent via sync helpers and
`do` function. Rather than choosing between synchronising every call you
can synchronise over a lexical/function scope with `do` or you can call
`sync` to perform a sync at any point for a particular contracts
instance (imagined to be request scoped). This is achieved by
maintaining a high-water mark form Burrow returns as well as
Postgres Vent notifications which can be compared.

Along the way cleaned up the contracts code in numerous ways:

- Better encapsulation of initialisation logic
- Use promise API of Burrow everywhere
- Better use of async and promises
- Improved vent helper (handles name reg results)
- Up-front failure on missing environment variables
- Await connections

Signed-off-by: Silas Davis <silas@monax.io>

And...

## Redaction

Add a `REDACTED` legal state available to agreement owners only that extends the cancellation flow. An owner may call `redact()` which will result in setting the legal state to `REDACTED` and emitting a Vent deletion event for the agreement only if:

- `msg.sender` is `owner`

and one of:

1. There are no parties (any legal state)
2. All parties have cancelled and `LegalState < FUFILLED`
2. Any party has cancelled and `LegalState = DRAFT`

It modifies the `CANCELLED` state machine so that in case 1 cancellation
will succeed.

Also adds vent synchronisation to all contract-controller functions

Signed-off-by: Silas Davis <silas@monax.io>

### CODE REVIEW

File: api/common/VentHelper.js [hunk 1]
```diff
@@ -26,43 +26,45 @@ class VentHelper {
   }
 
   waitForVent(result) {
-    if (result && result.height) {
+    const self = this;
+
+    return new Promise((resolve, reject) => {
+      if (!result && result.height === undefined) {
```
Comment (j-h-scheufen) [inline api/common/VentHelper.js]: Shouldn't this be an OR?

File: api/controllers/contracts-controller.js [hunk 2]
```diff
@@ -282,27 +287,33 @@ const isActiveArchetype = (archetypeAddress) => {
...
   return new Promise((resolve, reject) => {
-    archetype.isActive((err, data) => {
-      if (err || !data.raw) {
-        return reject(boom.badImplementation(`Failed to determine if archetype at ${archetypeAddress} is active: ${err}`));
-      }
-      log.info(`SUCCESS: Archetype at ${archetypeAddress} has been found to be ${data.raw[0] ? 'active' : 'inactive'}`);
-      return resolve(data.raw[0]);
-    });
+    archetype.isActive()
+      .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: isActive() is a 'view' function and does not require waitForVent as it's not a block transaction

File: api/controllers/contracts-controller.js [hunk 3]
```diff
@@ -282,27 +287,33 @@ const isActiveArchetype = (archetypeAddress) => {
...
   return new Promise((resolve, reject) => {
-    archetype.getAuthor((err, data) => {
-      if (err || !data.raw) {
-        return reject(boom.badImplementation(`Failed to get author of archetype at ${archetypeAddress}: ${err}`));
-      }
-      log.info(`SUCCESS: Retrieved archetype author for archetype at ${archetypeAddress}: ${data.raw[0]}`);
-      return resolve(data.raw[0]);
-    });
+    archetype.getAuthor()
+      .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: getAuthor() is a 'view' function. No waitForVent.

File: api/controllers/contracts-controller.js [hunk 4]
```diff
@@ -352,11 +363,13 @@ const setArchetypeSuccessor = (archetypeAddress, successorAddress, userAccount)
 const getArchetypeSuccessor = (archetypeAddress) => {
   log.debug(`REQUEST: Get successor for archetype at ${archetypeAddress}`);
   return new Promise((resolve, reject) => {
-    appManager.contracts['ArchetypeRegistry'].factory.getArchetypeSuccessor(archetypeAddress, (err, data) => {
-      if (err) return reject(boomify(err, `Failed to get successor for archetype at ${archetypeAddress}`));
-      log.info(`SUCCESS: Retrieved successor for archetype at ${archetypeAddress}`);
-      return resolve(data.raw[0]);
-    });
+    appManager.contracts['ArchetypeRegistry'].factory.getArchetypeSuccessor(archetypeAddress)
+      .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function. Remove waitForVent

File: api/controllers/contracts-controller.js [hunk 5]
```diff
@@ -692,8 +714,8 @@ const createUser = user => createUserInEcosystem(user, appManager.ecosystemAddre
 const getUserByIdAndEcosystem = (id, ecosystemAddress) => new Promise((resolve, reject) => {
   log.trace(`REQUEST: Get user by id: ${id} in ecosystem at ${ecosystemAddress}`);
   const ecosystem = getEcosystem(ecosystemAddress);
-  ecosystem
-    .getUserAccount(id)
+  ecosystem.getUserAccount(id)
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function. Remove waitForVent

File: api/controllers/contracts-controller.js [hunk 6]
```diff
@@ -1177,6 +1221,7 @@ const completeActivity = (actingUserAddress, activityInstanceId, dataMappingId =
   try {
     const bpmService = appManager.contracts['BpmService'];
     const piAddress = await bpmService.factory.getProcessInstanceForActivity(activityInstanceId)
+      .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function. Remove waitForVent

File: api/controllers/contracts-controller.js [hunk 7]
```diff
@@ -1230,14 +1275,14 @@ const completeActivity = (actingUserAddress, activityInstanceId, dataMappingId =
 const getModelAddressFromId = (modelId) => {
   log.debug(`REQUEST: Get model address for model id ${modelId}`);
   return new Promise((resolve, reject) => {
-    appManager.contracts['ProcessModelRepository'].factory.getModel(
-      hexFromString(modelId),
-      (error, data) => {
-        if (error || !data.raw) return reject(boom.badImplementation(`Failed to get address of model with id ${modelId}: ${error}`));
+    appManager.contracts['ProcessModelRepository'].factory.getModel(hexFromString(modelId))
+      .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 8]
```diff
@@ -1246,37 +1291,44 @@ const getProcessDefinitionAddress = (modelId, processId) => new Promise((resolve
   const modelIdHex = hexFromString(modelId);
   const processIdHex = hexFromString(processId);
   appManager.contracts['ProcessModelRepository']
-    .factory.getProcessDefinition(modelIdHex, processIdHex, (error, data) => {
-      if (error || !data.raw) {
+    .factory.getProcessDefinition(modelIdHex, processIdHex)
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 9]
```diff
@@ -1295,7 +1347,7 @@ const getStartActivity = processAddress => new Promise((resolve, reject) => {
   log.debug(`REQUEST: Get start activity id for process at address: ${processAddress}`);
   const processDefinition = getContract(process.env.API_ABI_DIRECTORY, BUNDLES.BPM_MODEL.contracts.PROCESS_DEFINITION, processAddress);
   processDefinition.getStartActivity()
-    .then(data => ventHelper.waitForVent(data))
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 10]
```diff
@@ -1311,44 +1363,56 @@ const getStartActivity = processAddress => new Promise((resolve, reject) => {
 const getProcessInstanceCount = () => new Promise((resolve, reject) => {
   log.debug('REQUEST: Get process instance count');
   appManager.contracts['BpmService']
-    .factory.getNumberOfProcessInstances((error, data) => {
-      if (error || !data.raw) return reject(boom.badImplementation(`Failed to get process instance count: ${error}`));
+    .factory.getNumberOfProcessInstances()
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 11]
```diff
@@ -1311,44 +1363,56 @@ const getStartActivity = processAddress => new Promise((resolve, reject) => {
...
 });
 
 const getProcessInstanceForActivity = activityInstanceId => new Promise((resolve, reject) => {
   log.debug(`REQUEST: Get process instance for activity ${activityInstanceId}`);
   appManager
     .contracts['BpmService']
-    .factory.getProcessInstanceForActivity(activityInstanceId, (error, data) => {
-      if (error || !data.raw) {
+    .factory.getProcessInstanceForActivity(activityInstanceId)
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 12]
```diff
@@ -1311,44 +1363,56 @@ const getStartActivity = processAddress => new Promise((resolve, reject) => {
...
-      log.info(`SUCCESS: Retrieved data mapping keys for process definition at ${processDefinition}, activity ${activityId} and direction ${direction}: ${JSON.stringify(keys)}`);
-      return resolve(keys);
-    }
-    log.info(`SUCCESS: No data mapping keys found for process definition at ${processDefinition}, activity ${activityId} and direction ${direction}`);
-    return resolve([]);
-  });
+    processDefinition.getInDataMappingKeys :
+    processDefinition.getOutDataMappingKeys;
+  countPromise(hexFromString(activityId))
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 13]
```diff
@@ -1391,26 +1455,31 @@ const getArchetypeProcesses = archAddress => new Promise((resolve, reject) => {
   let formation;
   let execution;
   appManager.contracts['ArchetypeRegistry']
-    .factory.getArchetypeData(archAddress, (err, data) => {
-      if (err || !data.raw) return reject(boom.badImplementation(`Failed to get archetype processes: ${err}`));
+    .factory.getArchetypeData(archAddress)
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 14]
```diff
@@ -1391,26 +1455,31 @@ const getArchetypeProcesses = archAddress => new Promise((resolve, reject) => {
...
 });
 
 const getActivityInstanceData = (piAddress, activityInstanceId) => new Promise((resolve, reject) => {
   log.debug(`REQUEST: Get activity instance data for activity id ${activityInstanceId} in process instance at address ${piAddress}`);
   appManager
     .contracts['BpmService']
-    .factory.getActivityInstanceData(piAddress, activityInstanceId, (err, data) => {
-      if (err || !data.raw) {
+    .factory.getActivityInstanceData(piAddress, activityInstanceId)
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/controllers/contracts-controller.js [hunk 15]
```diff
@@ -1423,17 +1492,21 @@ const getActivityInstanceData = (piAddress, activityInstanceId) => new Promise((
...
 });
 
 const getActiveAgreementData = agreementAddress => new Promise((resolve, reject) => {
   log.debug(`REQUEST: Get data for agreement at address ${agreementAddress}`);
   appManager
     .contracts['ActiveAgreementRegistry']
-    .factory.getActiveAgreementData(agreementAddress, (err, data) => {
-      if (err || !data.raw) {
+    .factory.getActiveAgreementData(agreementAddress)
+    .then(waitForVent)
```
Comment (j-h-scheufen) [inline api/controllers/contracts-controller.js]: View function

File: api/common/VentHelper.js (hunk 1 above)
Comment (silasdavis) [reply api/common/VentHelper.js]: oops yes

**Comment by silasdavis** [comment]
It was deliberate to put synchronisation on the view functions. Without knowing what you plan to do with the result of a read you cannot be sure it will be safe.

In the context of API this can break linearisation because we are effectively performing a join across two states (Burrow and Postgres) that are not synchronised. An advanced read from Burrow may cause you do the wrong thing in a SQL query.

Suppose we have three versioned values `A, B, C` that are written (by API from Burrow) at height 10. Suppose also that Burrow reads `A` at height 10 and then uses the value it reads to form a query `f` (we do this a lot) that also depends on other values. We can have the following trace:

```
time | burrow height | burrow write | burrow read |  postgres height | postgres write
1     | 9                     |  A1, B1        |                    |        7                 |
2     | 10                   |  A2, B2        | A2               |        8                 |
3     | 11                   |                    |                     |        9                 |  f(A2, B1) (from API)
4     | 12                   |                    |                     |        10               |  B2 (from vent)
```

In this trace we have writes in postgres `f(A2, B1), B2` but `A2` and `B2` was written in same block and so depend on each other. So what we should have had is `B2, f(A2, B2)`

Basically we have a dirty read (from postgres) in the context of our entire system. This is why combining these systems is hard we have to consider the consistency model across our composite database Burrow + Postgres.

To make the example more concrete suppose  `A` is the amount we pay to supplier `B`. Then suppose in block 10 we updated our supplier and the amount we pay them. Our advanced write of `A2` has gone into a query that has meant that we have set up a payment of the new amount to the old supplier. Because here `B` is coming from the value `B1` available to postgres.

If you are lucky then maybe `B` is used as a reference rather than a value and you are only temporarily inconsistent, but this is obviously not remotely guaranteed.

When a single request does not see its own writes reflected in the DB then it is more obviously pernicious but it can be just as bad in a read-modify cycle where a dirty read gets promoted to a write.

...

Having said all that... Burrow has a bug where it is not setting the height for read-only calls :-o

See: https://github.com/hyperledger/burrow/issues/1311 

So currently the vent helper will just be resolving these immediately. We should still keep the check in here and we can benefit from a fix in Burrow.

Another thing to say is that for simulated calls (used for view functions) you are not waiting for the _next_ block it will be served from a block already committed to state so you are not going to be waiting as long (if at all) as for writes.

**Comment by silasdavis** [comment]
Merging this since it is being used upstream - changes can be made before it gets released

**Pull request merged**